include/lsst/afw/image/TransmissionCurve.h

```diff
@@ -0,0 +1,300 @@
+     */
+    static std::shared_ptr<TransmissionCurve> multiply(
+        std::shared_ptr<TransmissionCurve> a,
+        std::shared_ptr<TransmissionCurve> b
+    );
+
+    /**
+     *  Return a view of a TransmissionCurve in a different coordinate system.
+     *
+     *  The transform is typically computed lazily (input points are
+     *  transformed and then the original TransmissionCurve is evaluated at
+     *  the transformed points).
+     *
+     *  @note This is wrapped as a bound method in Python (in C++ this
+     *        would be problematic due to the use of shared_ptr for base).
+     */
+    static std::shared_ptr<TransmissionCurve> transform(
+        std::shared_ptr<TransmissionCurve> base,
+        std::shared_ptr<geom::TransformPoint2ToPoint2> transform
+    );
+
+    // TransmissionCurve is not copyable.
+    TransmissionCurve(TransmissionCurve const &) = delete;
+
+    // TransmissionCurve is not movable.
+    TransmissionCurve(TransmissionCurve &&) = delete;
+
+    // TransmissionCurve is not copy-assignable.
+    TransmissionCurve & operator=(TransmissionCurve const &) = delete;
+
+    // TransmissionCurve is not move-assignable.
+    TransmissionCurve & operator=(TransmissionCurve &&) = delete;
+
+    // TransmissionCurve is polymorphic.
+    virtual ~TransmissionCurve() {}
+
+    /**
+     *  Return the bounds and spacing recommended for full evaluation.
+     *
+     *  Values within the returned min and max (inclusive) are guaranteed to
+     *  not yield NaN results when passed to sampleAt.
+     *
+     *  Min and/or max values may be infinite and size may be zero to indicate
+     *  an analytic curve with no wavelength bounds.
+     */
+    virtual SampleDef getNaturalSampling() const = 0;
+
+    /**
+     *  Return the throughput value that will be returned for wavelengths
+     *  below and above getNaturalSampling().min and .max (respectively).
+     *
+     *  Results may be NaN to indicate that the throughput is unknown, but must
+     *  otherwise be finite.
+     */
+    virtual std::pair<double,double> getThroughputAtBounds() const = 0;
+
+    //@{
+    /**
+     *  Evaluate the throughput at a position into a provided output array.
+     *
+     *  @param[in]  position     Spatial position at which to evaluate.
+     *  @param[in]  wavelengths  Wavelengths at which to evaluate.
+     *
+     *  @param[in,out]  out      Computed throughput values.  Must be pre-
+     *                           allocated to the same size as the wavelengths
+     *                           array.  May include NaN values if the
+     *                           requested wavelengths are beyond bounds
+     *                           reported by getWavelengthBounds(). Out-of-
+     *                           bounds wavelengths may also yield zero
+     *                           throughput values if throughput is known to
+     *                           be limited to those bounds.
```

Comment:
I'll document the exception when sizes are mismatched.

The order of output arguments may actually be something that merits an RFC to change or at least relax our guidelines; they're at odds with both the C++ standard library and (more relevant here) the NumPy convention for output array arguments.  I'll file that tomorrow.

Reply:
The argument for putting outputs first is that it allows for default values for input arguments. That said, I grant you it doesn't seem to be widely adopted and so can feel a bit unnatural. Mostly I try to avoid output arguments and modified arguments.

In any case, we do have the guideline so I think it's best to use it, unless you feel strongly enough about it to issue the RFC before merging the code.

Reply:
When the overloads are considered, this actually is a case where the output argument *is* optional (following the numpy pattern), and that's precisely why I like this order better.